js/stores/message_store.js

```diff
@@ -3,30 +3,47 @@ var assign = require('object-assign');
 var AppDispatcher = require('../dispatcher/app_dispatcher');
 var ActionTypes = require('../constants/constants').ActionTypes;
 
-var _messages = [];
+var _successfulMessages = [];
+var _pendingMessages = {};
 
 var MessageStore = assign({}, EventEmitter.prototype, {
 
   all: function() {
-    return _messages;
+    var messages = _successfulMessages.concat(this.pending());
+    var sorted = messages.sort(function(message1, message2) {
+      return message1.timestamp - message2.timestamp;
+    });
+
+    return sorted;
   },
 
   local: function() {
     var localMessages = [];
-    _messages.forEach(function(message) {
+    _successfulMessages.concat(this.pending()).forEach(function(message) {
       if (message.local === true) {
         localMessages.push(message);
       }
     });
 
     return localMessages;
+  },
+
+  pending: function() {
+    var messages = [];
+    for (var key in _pendingMessages) {
+      if (_pendingMessages.hasOwnProperty(key)) {
+        messages.push(_pendingMessages[key]);
+      }
+    }
+
+    return messages;
```

Comment:
you could use the utils/firebase `toArray` method (and rename the file to something more general than "firebase")